pkg/volume/photon_pd/attacher.go

```diff
@@ -0,0 +1,295 @@
+	}, nil
+}
+
+// Attaches the volume specified by the given spec to the given host.
+// On success, returns the device path where the device was attached on the
+// node.
+// Callers are responsible for retryinging on failure.
+// Callers are responsible for thread safety between concurrent attach and
+// detach operations.
+func (attacher *photonPersistentDiskAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
+	hostName := string(nodeName)
+	volumeSource, _, err := getVolumeSource(spec)
+	if err != nil {
+		glog.Errorf("Photon Controller attacher: Attach failed to get volume source")
+		return "", err
+	}
+
+	glog.V(4).Infof("Photon Controller: Attach disk called for host %s", hostName)
+
+	// Keeps concurrent attach operations to same host atomic
+	attachdetachMutex.LockKey(hostName)
+	defer attachdetachMutex.UnlockKey(hostName)
+
+	// TODO: if disk is already attached?
+	err = attacher.photonDisks.AttachDisk(volumeSource.PdID, nodeName)
+	if err != nil {
+		glog.Errorf("Error attaching volume %q: %+v", volumeSource.PdID, err)
+		return "", err
+	}
+
+	PdidWithNoHypens := strings.Replace(volumeSource.PdID, "-", "", -1)
+	return path.Join(diskByIDPath, diskPhotonPrefix+PdidWithNoHypens), nil
+}
+
+func (attacher *photonPersistentDiskAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.NodeName) (map[*volume.Spec]bool, error) {
+	volumesAttachedCheck := make(map[*volume.Spec]bool)
+	volumeSpecMap := make(map[string]*volume.Spec)
+	pdIDList := []string{}
+	for _, spec := range specs {
+		volumeSource, _, err := getVolumeSource(spec)
+		if err != nil {
+			glog.Errorf("Error getting volume (%q) source : %v", spec.Name(), err)
+			continue
+		}
+
+		pdIDList = append(pdIDList, volumeSource.PdID)
+		volumesAttachedCheck[spec] = true
+		volumeSpecMap[volumeSource.PdID] = spec
+	}
+	attachedResult, err := attacher.photonDisks.DisksAreAttached(pdIDList, nodeName)
+	if err != nil {
+		glog.Errorf(
+			"Error checking if volumes (%v) are attached to current node (%q). err=%v",
+			pdIDList, nodeName, err)
+		return volumesAttachedCheck, err
+	}
+
+	for pdID, attached := range attachedResult {
+		if !attached {
+			spec := volumeSpecMap[pdID]
+			volumesAttachedCheck[spec] = false
+			glog.V(2).Infof("VolumesAreAttached: check volume %q (specName: %q) is no longer attached", pdID, spec.Name())
+		}
+	}
+	return volumesAttachedCheck, nil
+}
+
+func (attacher *photonPersistentDiskAttacher) WaitForAttach(spec *volume.Spec, devicePath string, timeout time.Duration) (string, error) {
+	volumeSource, _, err := getVolumeSource(spec)
+	if err != nil {
+		glog.Errorf("Photon Controller attacher: WaitForAttach failed to get volume source")
+		return "", err
+	}
+
+	if devicePath == "" {
+		return "", fmt.Errorf("WaitForAttach failed for PD %s: devicePath is empty.", volumeSource.PdID)
+	}
+
+	// scan scsi path to discover the new disk
+	scsiHostScan()
+
+	ticker := time.NewTicker(checkSleepDuration)
```

Comment:
you should use `util.Poll` here

https://github.com/kubernetes/kubernetes/blob/master/pkg/util/wait/wait.go#L161